### PR TITLE
[WIP] Starting figuring out how to not remove wrappers

### PR DESCRIPTION
**Pull request by vsoch**
This is a work in progress to add an argument to not delete wrappers - I'm fairly new to the code base so it will also be a good practice in getting familiar. I'm currently trying to figure out all the points that we would want to check for the "don't do the delete" argument and then not do the cleanup, and I'm opening this PR (against my own branch) to get some help about where I need to be looking. I've added print debug statements, and then tested running a pipeline, but I don't see them being printed so I suspect I'm missing something!

Signed-off-by: Vanessa Sochat <vsochat@stanford.edu>

### CODE REVIEW

File: snakemake/executors.py [hunk 1]
```diff
@@ -897,12 +897,20 @@ def job_status(job):
 
             def job_status(job):
                 if os.path.exists(active_job.jobfinished):
-                    os.remove(active_job.jobfinished)
-                    os.remove(active_job.jobscript)
+                    if self.workflow.cleanup_wrappers:
```
Comment (johanneskoester) [inline snakemake/executors.py]: From your issue, this is actually not what you want to achieve I think. The jobscript is a shell script that is meant for submission to classical cluster systems.

I think what you meant to modify is the behavior here: https://github.com/snakemake/snakemake/blob/master/snakemake/script.py#L567
In other words, if your flag is false, that line shall not be executed.

And then, the flag should be called `--skip-cleanup-scripts`.

**Review by johanneskoester — CHANGES_REQUESTED** [review]
Goot start! See my comment below.

File: snakemake/executors.py (hunk 1 above)
Comment (vsoch) [reply snakemake/executors.py]: Ah, there it is! Thank you!

**Comment by vsoch** [comment]
@johanneskoester for your comment about where I opened the PR - I opened it intentionally here so that I could show you the diff between the current and my changes. I'll open the PR against the upstream (snakemake) when the PR is ready for review.

**Comment by vsoch** [comment]
@johanneskoester I'm still trying to figure out how to pass the parameter to the script. It looks like a script is part of a rule, and so far I've traced:

 - we start with the snakemake function, that generates a workflow (with a subsnakemake partial) and then we execute it.
 - the execution generates rules

```python
workflow._rules                                                                                             
Out[163]: 
OrderedDict([('a', <snakemake.rules.Rule at 0x7f7b6ff8d208>),
             ('b', <snakemake.rules.Rule at 0x7f7b6fcc1198>),
             ('c', <snakemake.rules.Rule at 0x7f7b6fcc1c88>)])
```
and also a JobScheduler (I think this isn't related as you mentioned so I won't get distracted here). Each rule has a script that starts out empty:

```python
 rule                                                                                
<snakemake.rules.Rule at 0x7f7b6fcc1198>
rule.script
```
The rules are given to a DAG and a persistence.Persistence object (not sure what these are), there's also a RuleInfo object that seems to work via passing around this "decorate" function everywhere. I can see that script is imported into workflow.py and it gets passed from a ruleinfo.script

```python
rule.script = ruleinfo.script
```
I've been interactively trying to walk through a dummy example and figure out how an input arg into snakemake even gets to the script.py file - two and a half hours now and I haven't had a clue, the logic of Snakemake is too complex for me to get my head around and I'm at a stopping point. How does the argument get passed to the script function, where we can designate for it not to be cleaned up?

**Comment by vsoch** [comment]
It looks like the wrapper also has a script created during it's init, but rule.wrapper is None for all my rules in this test case.

**Comment by johanneskoester** [comment]
Yeah, the route is complicated because of the way Snakemake implements it's DSL.

1. You need to add the additional parameter for the script function call [here](https://github.com/snakemake/snakemake/blob/master/snakemake/parser.py#L552), and [here](https://github.com/snakemake/snakemake/blob/master/snakemake/parser.py#L565).
2. You need to add the parameter to the run_wrapper [here](https://github.com/snakemake/snakemake/blob/master/snakemake/executors.py#L1889), and to the calls of the `run` function in that wrapper.
3. You need to add it where the arguments for the run_wrapper are prepared [here](https://github.com/snakemake/snakemake/blob/master/snakemake/executors.py#L353).

And yes, this is cumbersome and should be refactored at some point, into a parameter class that can be passed around as a single argument. It has grown over time, and we should've done the refactoring after the fifth parameter was added or so. If you want you can also do that now.

```
class RunParams:
    def __init__(self):
        self.rulename = None
        self.input = None
        ...

class WrapperParams(RunParams):
    ...
```

And so on.

**Comment by vsoch** [comment]
@johanneskoester I think we're missing one more spot for a rule maybe?
```
exception calling callback for <Future at 0x7f178fefe898 state=finished raised TypeError>
Traceback (most recent call last):
  File "/home/vanessa/Documents/Dropbox/Code/Python/snakemake/snakemake/executors.py", line 2035, in run_wrapper
    passed_shadow_dir
TypeError: __rule_b() takes 18 positional arguments but 19 were given
```

**Comment by vsoch** [comment]
Woohoo found it! It's under Run
```
class Run(RuleKeywordState):
    def __init__(self, snakefile, rulename, base_indent=0, dedent=0, root=True):
        super().__init__(snakefile, base_indent=base_indent, dedent=dedent, root=root)
        self.rulename = rulename
        self.content = 0

    def start(self):
        yield "@workflow.run"
        yield "\n"
        yield (
            "def __rule_{rulename}(input, output, params, wildcards, threads, "
            "resources, log, version, rule, conda_env, singularity_img, "
            "singularity_args, use_singularity, bench_record, jobid, "
            "is_shell, bench_iteration, cleanup_scripts, shadow_dir):".format(
                rulename=self.rulename
                if self.rulename is not None
                else self.snakefile.rulecount
            )
        )
```

**Comment by vsoch** [comment]
Opened an issue for the Run/Wrapper params idea https://github.com/snakemake/snakemake/issues/73 and this PR is no longer needed as it's finished and against upstream https://github.com/snakemake/snakemake/pulls. Closing.

**Pull request closed without merging**